waves-integration/src/main/scala/com/wavesplatform/dex/grpc/integration/clients/domain/portfolio/LookAheadPessimisticPortfolios.scala

```diff
@@ -15,46 +15,36 @@ import scala.util.chaining._
  */
 class LookAheadPessimisticPortfolios(orig: PessimisticPortfolios, maxConfirmedTransactions: Int) extends PessimisticPortfolios {
 
-  // TODO FifoSet
-  private val confirmedTxsEvictionQueue = new mutable.Queue[ByteString](maxConfirmedTransactions)
-  private val confirmedTxs = new mutable.HashSet[ByteString]
+  private val confirmedTxs = new mutable.LinkedHashSet[ByteString]
 
   override def getAggregated(address: Address): Map[Asset, Long] = orig.getAggregated(address)
 
   override def replaceWith(setTxs: Seq[PessimisticTransaction]): Set[Address] = {
     confirmedTxs.clear()
-    confirmedTxsEvictionQueue.clear()
     orig.replaceWith(setTxs)
   }
 
-  override def addPending(txs: Iterable[PessimisticTransaction]): Set[Address] =
-    orig.addPending(txs.filterNot(remove)) // Without unknown
+  override def addPending(txs: Iterable[PessimisticTransaction]): Set[Address] = {
+    val newTxs = txs.filterNot(tx => confirmedTxs.contains(tx.txId))
+    orig.addPending(newTxs) // Without unknown
+  }
 
   /**
    * @return (affected addresses, unknown transactions)
    */
-  override def processConfirmed(txIds: Iterable[ByteString]): (Set[Address], List[ByteString]) =
+  override def processConfirmed(txIds: Iterable[ByteString]): (Set[Address], List[ByteString]) = {
     // We don't filter, because a transaction can't be forged twice
-    orig.processConfirmed(txIds).tap { case (_, unknownTxIds) =>
-      unknownTxIds.foreach(put)
-    }
+    txIds.foreach(put)
+    orig.processConfirmed(txIds)
+  }
 
   override def removeFailed(txIds: Iterable[ByteString]): Set[Address] =
-    // txIds.foreach(remove) // a transaction can't be forged and failed both.
     orig.removeFailed(txIds)
 
   private def put(txId: ByteString): Unit = confirmedTxs.add(txId).tap { added =>
```

Comment:
i think we can simplify put...
```
private def put(txId: ByteString): Unit = confirmedTxs.add(txId).tap { added =>
    if (added && confirmedTxs.size > maxConfirmedTransactions)
      confirmedTxs.headOption.map(confirmedTxs.remove)
  }
  ```
  
  or even ( "size" operation should take O(1) )

```
  private def put(txId: ByteString): Unit = {
    confirmedTxs.add(txId)
    if (confirmedTxs.size > maxConfirmedTransactions)
      confirmedTxs.headOption.map(confirmedTxs.remove)
  }
  ```
  
and then we can remove import ```import scala.util.chaining._```